recipe/2024-06-private-bicep-registries.md

```diff
@@ -0,0 +1,269 @@
+
+During the recipe deployment bicep driver checks for secrets information for that container registry the recipe deploying is stored on,  and creates an oras auth client to authenticate the private bicep registries.
+
+
+e.g: creating a secret store with keys "username" and "password".
+
+```
+resource secretStore 'Applications.Core/secretStores@2023-10-01-preview' = {
+  name: 'acrsecrets'
+  properties:{
+    type: 'generic'
+    resource: 'registry-secrets/acr'
+    data: {
+      'username': {
+        value: '<username>'
+      }
+      'password': {
+        value: '<registry-password>'
+      }
+    }
+  }
+}
+```
+Update the recipe config in the environment to point to this secret resource id.
+```
+"recipeConfig": {
+  "terraform": {
+    ...
+  },
+  "bicep": {
+    "authentication":{
+      "test.azurecr.io":{
+        "secret": "/planes/radius/local/resourcegroups/default/providers/Applications.Core/secretStores/acrsecrets"
+      },
+    }
+  }
+  "env": {
+    ...
+  }
+},
+```
+
+#### Different container registry providers provide different authenticate methods to login.
+
+***Azure***:
+
+Azure provides multiple ways to authenticate ACR using ORAS, we can use different types of credentials depending on the setup and preferences. Here are the main types of credentials to use:
+
+#### Admin User Credentials:
+Users can enable the admin user and obtain these credentials from the Azure portal:
+```
+Username : The ACR admin username.
+Password : The ACR admin password.
+```
+
+#### Service Principal:
+```
+Username: The service principal's client ID.
+Password: The service principal's client secret.
+```
+
+#### Azure CLI Token:
+```
+Username: Use '00000000-0000-0000-0000-000000000000' (a placeholder indicating token-based authentication).
+Password: Use the token obtained from az acr login.
+```
+
+You can obtain the token using the Azure CLI:
+```
+az acr login --name <registry-name> --expose-token
+```
+***Github***:
+Github uses personal access token as password to auth into ghcr.
+```
+Username: Github username.
+Password: Personal access token.
+```
+
+***AWS***:
+```
+Username: AWS
+Password: output of `aws ecr get-login-password --region region`
+```
+
+### API design (if applicable)
+***Model changes***
+
+environment.tsp
+```diff
+@doc("Configuration for Recipes. Defines how each type of Recipe should be configured and run.")
+model RecipeConfigProperties {
+  @doc("Configuration for Terraform Recipes. Controls how Terraform plans and applies templates as part of Recipe deployment.")
+  terraform?: TerraformConfigProperties;
+
++  @doc(Configuration for Bicep Recipes. Controls how Bicep plans and applies templates as part of Recipe deployment.)
++  bicep?: BicepConfigProperties;
+
+  @doc("Environment variables injected during Terraform Recipe execution for the recipes in the environment.")
+  env?: EnvironmentVariables;
```

Comment:
Unrelated, but .... Is there a reason why `env` is part of the "recipe config" instead of the "terraform recipe config"

Reply:
fyi..updating comment on this to reflect env is part of Recipe config in upcoming PR